common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -179,133 +199,197 @@
-                LOGGER.info("load " + service.getSimpleName() + "[" + activateName + "] extension by class[" + extension
-                    .getName() + "]");
+    private S createExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes,
+                              Object[] args) {
+        Class<?> clazz = loadAllExtensionClass(loader).get(url);
+        if (clazz == null) {
+            throw new EnhancedServiceNotFoundException("Not Found");
+        }
+        try {
+            S instance = (S)EXTENSION_INSTANCES.get(clazz);
+            if (instance == null) {
+                S newInstance = initInstance(clazz, argTypes, args);
+                EXTENSION_INSTANCES.putIfAbsent(clazz, newInstance);
+                instance = newInstance;
             }
-            return result;
-        } catch (Throwable e) {
-            if (e instanceof EnhancedServiceNotFoundException) {
-                throw (EnhancedServiceNotFoundException)e;
-            } else {
-                throw new EnhancedServiceNotFoundException(
-                    "not found service provider for : " + service.getName() + " caused by " + ExceptionUtils
-                        .getFullStackTrace(e));
+            return instance;
+        } catch (Throwable t) {
+            throw new IllegalStateException("Extension instance(url: " + url + ", class: " +
+                    type + ")  could not be instantiated: " + t.getMessage(), t);
+        }
+    }
+
+    private S createNewExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes, Object[] args) {
+        Class<?> clazz = loadAllExtensionClass(loader).get(url);
+        if (clazz == null) {
+            throw new EnhancedServiceNotFoundException("Not Found");
+        }
+        try {
+            S newInstance = initInstance(clazz, argTypes, args);
+            return newInstance;
+        } catch (Throwable t) {
+            throw new IllegalStateException("Extension instance(url: " + url + ", class: " +
+                    type + ")  could not be instantiated: " + t.getMessage(), t);
+        }
+    }
+
+    private Map<ExtensionURL, Class<?>> loadAllExtensionClass(ClassLoader loader) {
+        Map<ExtensionURL, Class<?>> classes = extensionClasses.get();
+        if (classes == null) {
+            synchronized (extensionClasses) {
+                classes = extensionClasses.get();
+                if (classes == null) {
+                    classes = findAllExtensionClass(loader);
+                    extensionClasses.set(classes);
+                }
             }
         }
+        return classes;
     }
 
     @SuppressWarnings("rawtypes")
-    private static <S> List<Class> findAllExtensionClass(Class<S> service, String activateName, ClassLoader loader) {
-        List<Class> extensions = new ArrayList<>();
+    private Map<ExtensionURL, Class<?>> findAllExtensionClass(ClassLoader loader) {
+        Map<ExtensionURL, Class<?>> extensions = new HashMap<>();
```

Comment:
why not declare as TreeMap？

Reply:
yes ,treeMap is more appropriate.It has been changed

Reply:
Cause that treeMap uses comparable() to judge whether the extensionURLs are the same, but we need to judge by the hashCode() of extensionURLs, so here we can only declare as HashMap